bpf update: bcc remove linux kernel dep + devendor libbpf again, bpftrace 0.13.0 -> 0.14.0 + remove kernel dep, pahole 1.20 -> 1.22 + remove submodule, libbpf revert 0.6.0 -> 0.5.0 (unusable)

martinetd opened the pull request:
###### Motivation for this change

There are quite a few changes here, happy to split this up if it helps, the commits are all split correctly

- Remove kernel dependency
This is something I had on my mind for a while. It had been discussed in another PR/issue around last year but I can't find the link back, iirc the feedback was rather positive, but memory tends to be embellished so please comment again.
EDIT: found https://github.com/NixOS/nixpkgs/pull/118121#issuecomment-812991529 I think that was it -- in particular kheaders is loaded automatically if needed now so that point is no longer a problem

As far as I can test it's all positive, the only thing I didn't test were stap probes (usdt) but I don't see how the stap build time dependency could have any impact on this given the source so I assume it'll still work if anyone uses it.
bpftrace should also have no impact as far as I can see, I've used it a bit without any trouble.

- reverting libbpf to 0.5.0
~~Now that's something worth discussing.
The upgrade to 0.6.0 made bcc go back to the vendored libbpf because libbpf was unusable ( #139867 ) and nobody blinked at that because libbpf is a real pain with compatibily; I'm frankly tempted to just remove the libbpf dependency and have all its users use their submodules instead (as pointed out, it's always compiled as static anyway so size won't change)
The alternative I'd see would be for the libbpf package to ship its own uapi/linux/btf.h and have its users rely on this one instead of the one we provide in glibc's linux headers; that would remove 99% of the conflict we've seen with libbpf (it's really always that), but I don't know how to make a header preferred over another one to do this. We could explicitly remove the header from linux-headers? That could break other things, but they'd always get the updated libbpf one...
(and I mean come on, it's a release tag and there are things in libbpf's btf.h that haven't been merged in linux master! What if the patch upstream gets refused?!)~~
EDIT: no longer reverting this, instead installing linux/btf.h from libbpf to avoid that compatibility problem; this appears to work.

Meanwhile we really shouldn't ship a libbpf nobody can use, so reverting is the most sensible choice to me.

- bpftrace and pahole upgrade: these are straightforward and work without the rest, so could be split out if this drags on

- EDIT:  bpftrace fails to build with libbpf 0.6.0 due to a typeof misusage in libbpf, workaround by not relying on that typeof

- EDIT: add a nixos test for bpf stuff (bcc and libbpf, using a vm to run tests as root)

cc @Mic92 @r-burns @saschagrunert @jonringer @vcunat 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
  - I tried nixpkgs-review but it wanted to rebuild all kernels (which now depend on pahole) so stopped. I did rebuild bcc, bpftrace, pahole and libbpf with just about all variants of these patches, so I'm confident it works and that removing part of the PR is also OK.
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

martinetd committed:
bcc: move from linux-kernels packages to normal packages

bcc doesn't really need kernel itself, it just cares about module path.

It's actually better to use /run/booted-system/kernel-modules/lib/modules
for two reasons:
 - no need to rebuild bcc for each new kernel
 - can use a newer bcc with a booted kernel that doesn't match the current
   system

martinetd committed:
bpftrace: move from linux kernel packages to normal package

bpftrace doesn't actually need to be a kernel package anymore.
It used to require path to kernel sources, but we build our kernels
with IKHEADER and BTF so the currently running configuration can always
be found automatically without any patch

martinetd committed:
pahole: 1.20 -> 1.22

martinetd committed:
pahole: remove submodule

1.22 merged the EMBEDDED_LIBBPF patch back

martinetd committed:
bpftrace: 0.13.0 -> 0.14.0

martinetd committed:
bcc: remove submodule

reverting libbpf back to 0.5.0 fixes bcc compilation with our libbpf
This also shows this was a libbpf bug rather than a bcc problem...

martinetd committed:
nixos/tests: add bpf test

test bcc and bpftrace briefly

martinetd commented:
My original version had missed the nixos programs.bcc.enable, so figured I might as well add a test for this.

Took me some time to figure right commands that would exit immediately (most probes just run indefinitely..), but I'm happy with the result (I can remove the print() statements if that's a problem, but when running manually I think it's nice to see)
```
(finished: must succeed: syscount -d 1, in 13.43 seconds)
clock_gettime                 6
Tracing syscalls, printing top 10... Ctrl+C to quit.
[01:45:29]
SYSCALL                   COUNT
bpf                           3
ioctl                         2
epoll_pwait                   2
sendmsg                       1
epoll_ctl                     1
close                         1
pselect6                      1

Detaching...

machine: must succeed: bpftrace -l
(finished: must succeed: bpftrace -l, in 2.27 seconds)
machine: must succeed: bpftrace -e 'BEGIN { print("ok"); exit(); }'
(finished: must succeed: bpftrace -e 'BEGIN { print("ok"); exit(); }', in 1.07 seconds)
Attaching 1 probe...
ok



machine: must succeed: bpftrace -e 'tracepoint:syscalls:sys_enter_* { print(probe); exit(); }'
(finished: must succeed: bpftrace -e 'tracepoint:syscalls:sys_enter_* { print(probe); exit(); }', in 1.05 seconds)
Attaching 336 probes...
tracepoint:syscalls:sys_enter_bpf



machine: must succeed: bpftrace -e 'kprobe:schedule { print(probe); exit() }'
(finished: must succeed: bpftrace -e 'kprobe:schedule { print(probe); exit() }', in 0.50 seconds)
Attaching 1 probe...
kprobe:schedule


```

Mic92 commented on pkgs/development/tools/misc/pahole/default.nix:
```diff
@@ -1,20 +1,19 @@
-  nativeBuildInputs = [ cmake ];
-  buildInputs = [ elfutils zlib ];
+  nativeBuildInputs = [ cmake pkg-config ];
+  buildInputs = [ elfutils zlib libbpf ];
```
I had to remove the non-vendored version here because libbpf no longer build at some point: https://github.com/NixOS/nixpkgs/pull/139867#issue-1010741315

Mic92 replied:
How can we ensure we don't have to switch back and forth between vendoring and non-vendored version?

Mic92 commented on pkgs/os-specific/linux/libbpf/default.nix:
```diff
@@ -33,7 +33,7 @@ stdenv.mkDerivation rec {
     homepage    = "https://github.com/libbpf/libbpf";
     license     = with licenses; [ lgpl21 /* or */ bsd2 ];
-    maintainers = with maintainers; [ thoughtpolice vcunat saschagrunert ];
+    maintainers = with maintainers; [ thoughtpolice vcunat ];
```
```suggestion
    maintainers = with maintainers; [ thoughtpolice vcunat saschagrunert ];
```

martinetd replied:
Good point, I just blindly reverted the commit but will amend to keep @saschagrunert there.

Mic92 commented on nixos/tests/bpf.nix:
```diff
@@ -0,0 +1,25 @@
+
+    ## bpftrace
+    # list probes
+    machine.succeed("bpftrace -l")
```
This test can be added to `passthru.tests` of the bpftrace package.

martinetd replied:
I just found out about passthru.tests earlier today when trying to get OfBorg to run tests for me -- I think that's a great idea.
We do need this kind of tests for programs.bcc.enable though so should I just leave the bcc one there, or is the module simple enough (it's really just extra packages...) that this could go to bcc's passthru.tests instead as well? If that's easier to get coverage I'd be for moving it.

I'll do this as soon as I figure bcc's build first and check the exact syntax. Gotta afk a bit, might be a while - marking this PR draft for now.

martinetd replied:
I think I had misunderstood this comment - I wanted to try to move the commands as runCommand in bpftrace's passthru.test, but these don't run as root so we need to keep these here.

I've added nixosTests.bpf to bpftrace and bcc's passthru.test instead so now if either change this should automatically run. I assume that's what you had intended in the first place?

Mic92 replied:
yes.

martinetd commented:
(weird, I got mails but I see nothing on the web interface... I see your
other comments though..)

Jörg Thalheim wrote on Sat, Dec 04, 2021 at 12:45:04AM -0800:
> I had to remove the non-vendored version here because libbpf no longer
> build at some point:
> https://github.com/NixOS/nixpkgs/pull/139867#issue-1010741315 

Yes, I hate noticed -- that breakage wasn't bcc's but libbpf as I tried
to explain in the comment, basically libbpf 0.6.0 ships a btf.h that is
just totally unusable, and would have broken just about any project
using libbpf's btf code.

.... and !#@#@!, I had forgotten updating the checksum, so was just
rebuilding bcc with embedded libbpf :|

There seem to also be another problem, I'll look into it; but other
distros such as debian are also not using vendored libbpf so I think we
should encourage that way forward if it works.


> How can we ensure we don't have to switch back and forth between
> vendoring and non-vendored version?

Well, for the breakage that always comes back (incompatibility between
libbpf btf.h and linux header's), I've suggested shipping libbpf's
uapi/linux/btf.h instead of kernel header's -- but I'm not sure how to
do that.
Some extra tests when libbpf is upgraded might help?

Mic92 commented:
> (weird, I got mails but I see nothing on the web interface... I see your other comments though..) Jörg Thalheim wrote on Sat, 

I had my comments removed after re-reading your explanation.

Mic92 commented:
I might have an idea how to fix this btf.h priority, but currently I don't see which packages actually adds this header.
I see that `linuxPackages_*.kernel.dev` usually add this file but we no longer use the kernel package for bpftrace or bcc.

martinetd commented:
(comment removed -- ah, that makes sense, thanks for explaining :))

> I might have an idea how to fix this btf.h priority, but currently I don't see which packages actually adds this header.

iirc it comes from glibc somewhere, I found it once.. Let me have a look

martinetd commented:
Right, glibc:

`./pkgs/development/libraries/glibc/default.nix`
```
      if test -n "$linuxHeaders"; then
          # Include the Linux kernel headers in Glibc, except the `scsi'
          # subdirectory, which Glibc provides itself.
          (cd $dev/include && \
           ln -sv $(ls -d $linuxHeaders/include/* | grep -v scsi\$) .)
      fi
```

So we could just remove btf.h there and see what breaks?

Mic92 commented:
So glibc seems to have a `<linux/btf.h>` but not `<uapi/linux/btf.h>`.

Mic92 commented:
So you mean we need to install this header as well? https://github.com/libbpf/libbpf/blob/master/include/uapi/linux/btf.h

martinetd commented:
That's the one. uapi is the way the linux kernel marks headers to be installed in /usr/include on "normal" systems, so libbpf put it there.

This header defines the enum (used to be defines and got converted to enum)  with e.g. BTF_KIND_FLOAT that was a problem last time when it got added.

Mic92 commented:
So as long as we just install this header file and build bpf packages without other linux kernel headers, we should be good to go.

martinetd commented:
I keep missing your last reply when I send mine so I'll recap once and hope this is clearer, we have (at least) three btf.h headers:

- the linux kernel provides (path in kernel sources) `./include/uapi/linux/btf.h` which is installed through glibc, that provides the enum with `BTF_KIND_FLOAT` - it's normally installed in `<include path>/linux/btf.h`
- libbpf has https://github.com/libbpf/libbpf/blob/master/include/uapi/linux/btf.h which is the same, but in a version that'll always work with itself
- libbpf also provides and installs https://github.com/libbpf/libbpf/blob/master/src/btf.h (usually installed in `<include path>/bpf/btf.h`) , which uses values such as `BTF_KIND_FLOAT`


> So as long as we just install this header file and build bpf packages without other linux kernel headers, we should be good to go.

Yes, I think that'll fix most of the problems we've had with libbpf over the past few years

martinetd commented:
bcc build failure:

hm, this is really weird. I've entered a build shell with `nix-shell -A bcc`, ran unpackPhase cmakeConfigurePhase and... there's just no include path for libbpf?

```
$ env | grep bpf
nativeBuildInputs=/nix/store/k0z9n599k02hab8qjjp3ljw065iwjcvg-python3-3.9.6 /nix/store/ja2c6vmg2xz40v6xxz0pjjvls3f1wx59-hook /nix/store/vv743igvmpgr25p0vihyf4pms08hi6sr-hook /nix/store/zmb027dm68kgpx5pg1xgfbq11qxybm13-python-remove-tests-dir-hook /nix/store/fivxgjz8rplffmzijnw8sgqn9ynidrkl-python3.9-setuptools-57.2.0 /nix/store/hr5q7snw3rjmvgw88q2grisn32cmmr0z-python-catch-conflicts-hook /nix/store/ky0kj7cgh8k4rmf5bq4cw16ldc754vjr-python-remove-bin-bytecode-hook /nix/store/k2na6g7ykmc3ck8vmyj3vhx4pcw4i3n9-python-imports-check-hook.sh /nix/store/sdhnq0ngx5kfz55q2lrfdc5jjrpri3bq-python-namespaces-hook.sh /nix/store/6rnp1psxb8lhyrgfaif9hvs94anvs3hg-hook /nix/store/q2lj6pxd6sn75fa9pjx9x3jpgxnr6zld-cmake-3.21.2 /nix/store/p6kqbla658ksalqwapplsas7qfrk16xb-flex-2.6.4 /nix/store/6kc28ix1fng215s3dx88fi94nbjfp976-bison-3.8.2 /nix/store/d1sqpma11wgqjymffik7hyx7l56f666r-llvm-7.1.0-dev /nix/store/bm2ddhnxq8kpwlapvsd7v6vdwcgi5bnn-libbpf-0.5.0
```

`NIX_CFLAGS_COMPILE` doesn't include the libbpf installdir/include, similarly for libs, and there's no PKG_CONFIG_PATH or equivalent either.

I don't see how this could have ever worked, I think I'm missing something and nix-shell isn't the way to test builds anymore - what am I missing?

martinetd commented:
force pushed with:
- fixed bcc build (updated hash, put libbpf in the right place and cmake flags) - I've also checked by the way that the failure with 0.6.0 is what I thought it was, I'm somewhat confident it should build with the right linux/btf.h
- re-added @saschagrunert to libbpf maintainers

I need to afk now, will move tests to passthru.tests after dinner

martinetd committed:
libbpf: install linux/btf.h

This installs the linux/btf.h header in libbpf, which is necessarily
compatible with bpf/btf.h
Note that we also ship the linux header's linux/btf.h with glibc,
but explicit buildInputs should have priority over default headers so
hopefully this one will always be used.

martinetd commented:
So i've tried just installing libbpf's linux/btf.h and hope the right one is first in search path -- and was lucky on my local system.

I could indeed build both bcc and pahole with libbpf 0.6.0 and the following patch:
```
From 6961308155bf978882aba5705baa9043a7577494 Mon Sep 17 00:00:00 2001
From: Dominique Martinet <asmadeus@codewreck.org>
Date: Sat, 4 Dec 2021 20:25:09 +0900
Subject: [PATCH] libbpf: install linux/btf.h

This installs the linux/btf.h header in libbpf, which is necessarily
compatible with bpf/btf.h
Note that we also ship the linux header's linux/btf.h with glibc, so
as long as both exist there is no guarantee the right one will be
used

diff --git a/pkgs/os-specific/linux/libbpf/default.nix b/pkgs/os-specific/linux/libbpf/default.nix
index 8c6b5d0c553c..ba765b4f92e2 100644
--- a/pkgs/os-specific/linux/libbpf/default.nix
+++ b/pkgs/os-specific/linux/libbpf/default.nix
@@ -23,6 +23,11 @@ stdenv.mkDerivation rec {
   enableParallelBuilding = true;
   makeFlags = [ "PREFIX=$(out)" ];
 
+  postInstall = ''
+    # install linux's libbpf-compatible linux/btf.h
+    install -Dm444 ../include/uapi/linux/btf.h -t $out/include/linux
+  '';
+
   # FIXME: Multi-output requires some fixes to the way the pkg-config file is
   # constructed (it gets put in $out instead of $dev for some reason, with
   # improper paths embedded). Don't enable it for now.

```

I don't think we should do it unless we remove it from glibc's though, what do you think?

Mic92 commented:
> So i've tried just installing libbpf's linux/btf.h and hope the right one is first in search path -- and was lucky on my local system.
> 
> I could indeed build both bcc and pahole with libbpf 0.6.0 and the following patch:
> 
> ```
> From 6961308155bf978882aba5705baa9043a7577494 Mon Sep 17 00:00:00 2001
> From: Dominique Martinet <asmadeus@codewreck.org>
> Date: Sat, 4 Dec 2021 20:25:09 +0900
> Subject: [PATCH] libbpf: install linux/btf.h
> 
> This installs the linux/btf.h header in libbpf, which is necessarily
> compatible with bpf/btf.h
> Note that we also ship the linux header's linux/btf.h with glibc, so
> as long as both exist there is no guarantee the right one will be
> used
> 
> diff --git a/pkgs/os-specific/linux/libbpf/default.nix b/pkgs/os-specific/linux/libbpf/default.nix
> index 8c6b5d0c553c..ba765b4f92e2 100644
> --- a/pkgs/os-specific/linux/libbpf/default.nix
> +++ b/pkgs/os-specific/linux/libbpf/default.nix
> @@ -23,6 +23,11 @@ stdenv.mkDerivation rec {
>    enableParallelBuilding = true;
>    makeFlags = [ "PREFIX=$(out)" ];
>  
> +  postInstall = ''
> +    # install linux's libbpf-compatible linux/btf.h
> +    install -Dm444 ../include/uapi/linux/btf.h -t $out/include/linux
> +  '';
> +
>    # FIXME: Multi-output requires some fixes to the way the pkg-config file is
>    # constructed (it gets put in $out instead of $dev for some reason, with
>    # improper paths embedded). Don't enable it for now.
> ```
> 
> I don't think we should do it unless we remove it from glibc's though, what do you think?

I think `buildInputs` should have priority over glibc's own header in our cc-wrapper, so we would be safe unless someone adds `glibc` to `buildInputs`.

martinetd commented:
> I think buildInputs should have priority over glibc's own header in our cc-wrapper, so we would be safe unless someone adds glibc to buildInputs.

Fair enough. Pushing this patch then and removing draft here, let's see how this goes :)

martinetd committed:
maintainers: add martinetd

will add myself to bpf-related packages over the next few commits

martinetd commented:
Gah, why was I convinced that if bcc and pahole worked bpftrace also would... bpftrace fails to build with libbpf 0.6.0:
```
[ 57%] Building CXX object src/ast/CMakeFiles/ast.dir/async_event_types.cpp.o
In file included from /build/source/src/btf.cpp:23:
/build/source/src/btf.cpp: In member function 'std::string bpftrace::BTF::c_def(const std::unordered_set<std::__cxx11::basic_string<char> >&) const':
/build/source/src/btf.cpp:228:10: error: 'typeof' was not declared in this scope; did you mean 'type_of'?
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^~~~~~~~~~~~~
/build/source/src/btf.cpp:228:10: error: expected primary-expression before ')' token
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^~~~~~~~~~~~~
/build/source/src/btf.cpp:228:10: error: '__builtin_types_compatible_p' was not declared in this scope
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^~~~~~~~~~~~~
/build/source/src/btf.cpp:228:10: error: expression list treated as compound expression in functional cast [-fpermissive]
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^~~~~~~~~~~~~
/build/source/src/btf.cpp:228:10: error: expected primary-expression before 'void'
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^~~~~~~~~~~~~
/build/source/src/btf.cpp:228:10: error: invalid conversion from 'void*' to 'const btf*' [-fpermissive]
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^
      |          |
      |          void*
/nix/store/0f1qz331r4wb03xjs2jcsbnw9xsdn2qk-libbpf-0.6.0/include/bpf/btf.h:289:72: note:   initializing argument 1 of 'btf_dump* btf_dump__new_deprecated(const btf*, const btf_ext*, const btf_dump_opts*, btf_dump_printf_fn_t)'
  289 | LIBBPF_API struct btf_dump *btf_dump__new_deprecated(const struct btf *btf,
      |                                                      ~~~~~~~~~~~~~~~~~~^~~
/build/source/src/btf.cpp:228:10: error: invalid conversion from 'void*' to 'const btf_ext*' [-fpermissive]
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^
      |          |
      |          void*
/nix/store/0f1qz331r4wb03xjs2jcsbnw9xsdn2qk-libbpf-0.6.0/include/bpf/btf.h:290:34: note:   initializing argument 2 of 'btf_dump* btf_dump__new_deprecated(const btf*, const btf_ext*, const btf_dump_opts*, btf_dump_printf_fn_t)'
  290 |            const struct btf_ext *btf_ext,
      |            ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/build/source/src/btf.cpp:228:10: error: invalid conversion from 'void*' to 'const btf_dump_opts*' [-fpermissive]
  228 |   dump = btf_dump__new(btf, nullptr, &opts, dump_printf);
      |          ^
      |          |
      |          void*
/nix/store/0f1qz331r4wb03xjs2jcsbnw9xsdn2qk-libbpf-0.6.0/include/bpf/btf.h:291:40: note:   initializing argument 3 of 'btf_dump* btf_dump__new_deprecated(const btf*, const btf_ext*, const btf_dump_opts*, btf_dump_printf_fn_t)'
  291 |            const struct btf_dump_opts *opts,
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
```

libbpf uses typeof to chose between two versions of btf_dump__new and apparently that's not valid for c++ code, I'll work around it in bpftrace for now...

martinetd committed:
bpftrace: fix build with libbpf 0.6.0

libbpf 0.6.0 relies on typeof in bpf/btf.h to pick the right version of
btf_dump__new() but that's not valid c++.
see https://github.com/iovisor/bpftrace/issues/2068

martinetd commented:
And we're probably good again... My tests are waiting for kernel rebuild to finish now, I'll post the result tomorrow if I'm faster than ofBorg (doubtful)

So that workaround isn't 100% fool-proof, but this one is a bona fide bug so I guess we can just deal with these from time to time.. I've commented on https://github.com/iovisor/bpftrace/issues/2068 asking what their plan is for this bug, they just pinned libbpf 0.5.0 in their CI for now

Mic92 commented:
Thanks. This should make bcc a lot easier to use.

arianvp commented:
Hmm how is this supposed to work? I can't make bpftrace find kernel headers.  Tracing back the error message leads me to `get_kernel_dirs()` which we used to patch before this PR was merged

```
$ sudo bpftrace imds.bt
WARNING: Could not find kernel headers in  or . To specify a particular path to kernel headers, set the env variables BPFTRACE_KERNEL_SOURCE and, optionally, BPFTRACE_KERNEL_BUILD if the kernel was built in a different directory than its source. To create kernel headers run 'modprobe kheaders', which will createa tar file at /sys/kernel/kheaders.tar.xz
definitions.h:2:10: fatal error: 'net/sock.h' file not found
```

arianvp commented:
Ah! `programs.bcc.enable = true` fixed it

martinetd commented:
FWIW I'm not sure programs.bcc.enable change much here?

This should work because bcc is built with modules path of /run/booted-system/kernel-modules/lib/modules, which should contain kheaders.ko (nixos kernel is built with CONFIG_IKHEADES=m), and bpftrace until 0.20.4 (excluded) would extract these when required so it will find net/sock.h there... 
Unfortunately because of security problems of using /tmp unsafely they removed the extraction in the stable branch of 0.20.4 (something else might come back in 0.21), so you'll just get a fat warning explaining how to extract it manually with the latest version, and I don't see how programs.bcc would help.

In this particular case, with BTF (CONFIG_DEBUG_INFO_BTF=y on nixos kernel) you actually don't even need that include, bpftrace can resolve `struct sock` without it, so just removing the include works with the latest version (the warning is however still displayed)

I see https://github.com/bpftrace/bpftrace/pull/3156 says the warning should only be displayed if headers are actually needed but this obviously doesn't seem to work, I'll look a bit more into it later....

arianvp commented:
So i can only reproduce this when i install BPFTrace using nix shell nixpkgs#bpftrace as opposed to systemPackages. Not sure why that makes a difference

martinetd commented:
I think that would be the version, I'd wager you get the older 0.20.3 via the system.
The new 0.20.4 (that you get with nix shell) will work if you just remove the `#include` line, it's no longer required with BTF present

arianvp commented:
Thank you!

arianvp commented:
Gotcha.  I updated now to `0.20.4` on all my machines and it indeed works without the headers. However the pesky `WARNING` remains.

```
Jun 11 11:22:18 utm imdstrace-start[1360]: WARNING: Could not find kernel headers in  or . To specify a particular path to kernel headers, set the env variables BPFTRACE_KERNEL_SOURCE and, optionally, BPFTRACE_KERNEL_BUILD if the kernel was built in a different directory than its source. To create kernel headers run 'modprobe kheaders', which will create a tar file at /sys/kernel/kheaders.tar.xz
Jun 11 11:22:18 utm imdstrace-start[1361]: {"type": "attached_probes", "data": {"probes": 2}}
```

Also it's kinda annoying  because basically all examples in https://github.com/bpftrace/bpftrace/tree/master/tools include headers and non of them now work out of the box on NixOS. They all crash

arianvp commented:
e.g. all the `.bt` programs in path don't work:

```
$ sudo tcpaccept.bt 
[sudo] password for arian: 
WARNING: Could not find kernel headers in  or . To specify a particular path to kernel headers, set the env variables BPFTRACE_KERNEL_SOURCE and, optionally, BPFTRACE_KERNEL_BUILD if the kernel was built in a different directory than its source. To create kernel headers run 'modprobe kheaders', which will create a tar file at /sys/kernel/kheaders.tar.xz
definitions.h:6:10: fatal error: 'sys/socket.h' file not found
```

arianvp commented:
I think we need to somehow include `glibc.dev` as a default include path for these to work.

(By the way; sorry for hijacking this PR to leave my ramblings. Happy to open up a new tracking issue for this)
